File: 0752-open-the-lock/0752-open-the-lock.py
class Solution:
    def neighbors(self, node: List[str]) -> List[str]:
        ans = []
        for i in range(4):
            num = int(node[i])
            for change in [-1, 1]:
                x = (num + change) % 10
                ans.append(node[:i] + str(x) + node[i + 1:])
        return ans
    
    def openLock(self, deadends: List[str], target: str) -> int:
        if "0000" in deadends:
            return -1
        
        queue = deque([("0000", 0)])
        seen = set(deadends)
        seen.add("0000")
        
        while queue:
            node, steps = queue.popleft()
            if node == target:
                return steps
            
            for neighbor in self.neighbors(node):
                if neighbor not in seen:
                    seen.add(neighbor)
                    queue.append((neighbor, steps + 1))
        return -1

File: 0752-open-the-lock/0752-open-the-lock.cpp
class Solution {
    vector<string> neighbors(string& node){
        vector<string> node_bfs;
        for (int i = 0; i < 4; ++i){
            int num = int(node[i] - '0');
            for (int change : {-1, 1}){
                int x = (num + change + 10) % 10;
                node_bfs.push_back(node.substr(0, i) + (char)(x + 48) + node.substr(i+1));
            }
        }
        
        return node_bfs;
    }
    
public:
    int openLock(vector<string>& deadends, string target) {
        if (find(deadends.begin(), deadends.end(), "0000") != deadends.end()) return -1;
        
        queue<pair<string, int>> queue;
        queue.push({"0000", 0});
        unordered_set<string> seen(deadends.begin(), deadends.end());
        seen.insert("0000");
        
        while (!queue.empty()){
            auto it = queue.front();
            queue.pop();
            string node = it.first;
            int steps = it.second;
            if (node == target) return steps;
            for (string& neighbor : neighbors(node)){
                if (seen.find(neighbor) == seen.end()){
                    seen.insert(neighbor);
                    queue.push({neighbor, steps + 1});
                }
            }
        }
        
        return -1;
    }
};